pkg/controller/projectreference/projectreference_adapter.go

```diff
@@ -0,0 +1,341 @@
+
+	}
+	return projectClaim, nil
+}
+
+// updateProjectID updates the ProjectReference with a unique ID for the ProjectID
+func (r *ReferenceAdapter) updateProjectID() error {
+	guid := uuid.New().String()
+	hashing := sha1.New()
+	hashing.Write([]byte(guid))
+	uuidsum := fmt.Sprintf("%x", hashing.Sum(nil))
+	shortuuid := uuidsum[0:26]
+	r.projectReference.Spec.GCPProjectID = "osd-" + shortuuid
+	return r.kubeClient.Update(context.TODO(), r.projectReference)
+}
+
+// updateProjectID updates the ProjectReference with a unique ID for the ProjectID
+func (r *ReferenceAdapter) clearProjectID() error {
+	r.projectReference.Spec.GCPProjectID = ""
+	return r.kubeClient.Update(context.TODO(), r.projectReference)
+}
+
+// checkRequirements checks that region is supported
+func (r *ReferenceAdapter) checkRequirements() error {
+	if _, ok := supportedRegions[r.projectClaim.Spec.Region]; !ok {
+		return operrors.ErrRegionNotSupported
+	}
+	return nil
+}
+
+func (r *ReferenceAdapter) createProject(parentFolderID string) error {
+	// Get existing projects
+	projects, err := r.gcpClient.ListProjects()
+	if err != nil {
+		return err
+	}
+
+	projectMap := convertProjectsToMap(projects)
+
+	if project, ok := projectMap[r.projectReference.Spec.GCPProjectID]; ok {
+		switch project.LifecycleState {
+		case "ACTIVE":
+			r.logger.Info("Project lifecycleState == ACTIVE")
+			return nil
+		case "DELETE_REQUESTED":
+			return operrors.ErrInactiveProject
+		default:
+			r.logger.Error(operrors.ErrUnexpectedLifecycleState, "Unexpected LifecycleState", project.LifecycleState)
+			return operrors.ErrUnexpectedLifecycleState
+
+		}
+	}
+
+	r.logger.Info("Creating Project")
+	// If we cannot create the project clear the projectID from spec so we can try again with another unique key
+	_, err = r.gcpClient.CreateProject(parentFolderID)
+	if err != nil {
+		r.logger.Error(err, "could not create project", "Parent Folder ID", parentFolderID, "Requested Project ID", r.projectReference.Spec.GCPProjectID)
+		r.logger.Info("Clearing gcpProjectID from ProjectReferenceSpec")
+		err = r.clearProjectID()
+		if err != nil {
+			return err
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReferenceAdapter) configureAPIS() error {
+	billingAccount, err := util.GetBillingAccountFromSecret(r.kubeClient, operatorNamespace, orgGcpSecretName)
+	if err != nil {
+		r.logger.Error(err, "Could not get org billingAccount from secret", "Secret Name", orgGcpSecretName, "Operator Namespace", operatorNamespace)
+		return err
+	}
+
+	r.logger.Info("Enabling Billing API")
+	err = r.gcpClient.EnableAPI(r.projectReference.Spec.GCPProjectID, "cloudbilling.googleapis.com")
+	if err != nil {
+		r.logger.Error(err, fmt.Sprintf("Error enabling %s api for project %s", "cloudbilling.googleapis.com", r.projectReference.Spec.GCPProjectID))
+		return err
+	}
+
+	r.logger.Info("Linking Cloud Billing Account")
+	// https://groups.google.com/forum/#!topic/gce-discussion/K_x9E0VIckk
```

Comment:
The comment can be dropped, this works without issues if proper perms are given to the service account.

Reply:
See https://github.com/rafael-azevedo/gcp-project-operator-1/pull/1